### PR TITLE
remove unnecessary -Wno= flags

### PR DESCRIPTION
**Pull request by anderspapitto**

### CODE REVIEW

**Comment by colesbury** [comment]
Wmaybe-uninitialized is pretty flaky, especially since we use a variety of GCC versions.

**Comment by ezyang** [comment]
Maybe we should run Werror only on a single compiler configuration; either the latest (less likely to have bugs) or the compiler that all the devs use (so it's easy to repro locally).

**Comment by anderspapitto** [comment]
@goldsborough thoughts on the comments of @colesbury and @ezyang ?

**Comment by goldsborough** [comment]
@colesbury do you have examples of the flakiness? I quenched this warning in all of ATen/PyTorch a couple months ago and only had to add `(void)x` to a few places. The primary issue was when code was `#ifdef`-ed out. I think quenching the edge cases with `(void)x` would be fine. I remember places in the autograd where integers were not initialized to zero (usually member variables) and we're just kind of lucky they always ended up being zero

I personally think keeping warnings on in all CI is better because otherwise they'll just stay silent until someone builds with a particular configuration (e.g. on the FAIR cluster or on macOs) on their local machine, and then we'll have to fix the warning spew on those configurations anyway

**Comment by anderspapitto** [comment]
I consider this diff ready to merge

**Comment by colesbury** [comment]
It doesn't work well with tagged enums like in `Scalar.h`. The current code doesn't issue a warning for very subtle reasons. Slight changes to the code (like moving a method inline) will break the build.

It also fails with things like https://github.com/pytorch/pytorch/blob/5e84403d5fc92ad2ed594b86a436c571dac55d30/aten/src/THCUNN/generic/BatchNormalization.cu#L28

I'm not sure what `(void)x` has to do with `-Wmaybe-uninitialized`. `maybe-uninitialized` doesn't check for unused variables.

The heuristics used depend on compiler version. For example, locally with GCC 4.8.1 I see warnings about that BatchNormalization line, but later versions of GCC are slightly smarter. The reverse is often true too.

We work really hard to avoid flakiness in unit tests. We should take the same approach when dealing with compiler warnings, particularly when they are treated as errors.

The `-Wuninitialized` warning is **much** better. We have that on and should keep that on.

**Comment by goldsborough** [comment]
Ok, sounds like good reasons to disable the warning. My brain also confused `maybe-uninitialized` with `maybe-unused`, the `(void)x` argument doesn't apply -- my bad.

@anderspapitto please disable `-Wmaybe-uninitialized`

**Comment by anderspapitto** [comment]
reintroduced -Wno-maybe-uninitialized

**Comment by ezyang** [comment]
@goldsborough @colesbury  maybe-uninitialized is twinging on @resistor's PR at #9640 so it sounds like it is currently disabled for all code.

File: torch/CMakeLists.txt [hunk 1]
```diff
@@ -230,6 +230,14 @@ else()
     # warnings, see https://bugs.llvm.org/show_bug.cgi?id=21629
     -Wno-missing-braces
     )
+
+  if(NOT APPLE)
+    target_compile_options(torch PRIVATE
+      # Considered to be flaky.  See the discussion at
+      # https://github.com/pytorch/pytorch/pull/9608
+      -Wno-maybe-uninitialized)
```
Comment (ezyang) [inline torch/CMakeLists.txt]: Why is this condition on NOT APPLE?

**Review by ezyang — APPROVED** [review]

File: torch/CMakeLists.txt (hunk 1 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: the compiler on the macos CI machines doesn't recognize the existence of this option
